client/state/sites/reducer.js

```diff
@@ -98,18 +99,38 @@ export function items( state = {}, action ) {
 			return state;
 		}
 
-		case SITE_RECEIVE: {
-			const site = pick( action.site, VALID_SITE_KEYS );
-			return Object.assign( {}, state, {
-				[ site.ID ]: site
-			} );
-		}
-
+		case SITE_RECEIVE:
 		case SITES_RECEIVE:
-			return action.sites.reduce( ( memo, site ) => {
-				memo[ site.ID ] = pick( site, VALID_SITE_KEYS );
+		case SITES_UPDATE:
+			// Normalize incoming site(s) to array
+			const sites = action.site ? [ action.site ] : action.sites;
+
+			// SITES_RECEIVE occurs when we receive the entire set of user
+			// sites (replace existing state). Otherwise merge into state.
+			const initialMemo = SITES_RECEIVE === action.type ? {} : state;
+
+			return reduce( sites, ( memo, site ) => {
+				// If we're not already tracking the site upon an update, don't
+				// merge into state (TODO: Consider dropping condition once
+				// sites-list abolished)
+				if ( SITES_UPDATE === action.type && ! memo[ site.ID ] ) {
+					return memo;
+				}
+
+				// Bypass if site object hasn't change
+				const transformedSite = pick( site, VALID_SITE_KEYS );
+				if ( isEqual( memo[ site.ID ], transformedSite ) ) {
+					return memo;
+				}
+
+				// Avoid mutating state
+				if ( memo === state ) {
+					memo = { ...state };
+				}
+
+				memo[ site.ID ] = transformedSite;
 				return memo;
-			}, {} );
+			}, initialMemo );
```

Comment:
>`accumulator` or `initialValue` is pretty common. In this case though, what about `lastState`?
  
Oof, I thought I'd read this as `initialState` when I'd updated in rebased 9cc2607. Though `lastState` is not entirely accurate, since it's not so much what the previous state is, but the basis for what the next state will be. Maybe something like `initialNextState`? Quite a mouthful.